### DISCUSSION TITLE
Add minimal DoH support to tpws

### DISCUSSION THREAD

**Discussion by NewUse** [bol-van/zapret#447] [Ideas]
Hi all, it seems possible to check the request header of http request and rewrite destination IP address by provided by trusted DoH server instead of untrusted IP provided by ISPs DNS.

So it should look like --doh=1.1.1.1,1.0.0.1
and some check option to  be able to check DoH certificate before setting up this option or may be this check should be added to blockcheck.sh

**Comment by bol-van** [comment]
It requires connection delay to the moment when hostname is known. Not easy coding but possible.
Also it requires heavy curl/crypto libs linked to tpws. It will ruin small tpws binary
Not worth doing

**Reply by NewUse** [reply]
In case of sizing of tpws dependencies will grow significantly I think that this feature is useless -- there are doh-proxies but size is so big that it is impossible to run it on old 4MB flash devices, for other cases there are nfqws + https-dns-proxy that are more suitable  for modern devices with enough flash size.

**Comment by testing765** [comment]
Вот кстати подозреваю что отсутствие поддержки DoH сводит полезность blockcheck на 0
Например сейчас проверил blockcheck для rutor.info и мне выдало:
```
- checking winws --wf-l3=ipv4 --wf-tcp=443 --dpi-desync=disorder2
curl: (60) SSL certificate problem: certificate has expired
More details here: https://curl.se/docs/sslcerts.html
```
и вообще в конце, в summary такой результат:
```
* SUMMARY
ipv4 rutor.info curl_test_https_tls12 : winws not working
ipv4 rutor.info curl_test_http3 : winws not working
```
но если я запускаю winws с этим же **--wf-tcp=443 --dpi-desync=disorder2** , то в хроме с включенным DoH все работает отлично и серт действительный

**Reply by bol-van** [reply]
Блокчек в начале проверяет DNS и орет, если идет подмена.
Не он, а вы должны решить проблему с DNS.

Если же DNS не подменяется, а проблема лишь в том, что системный DNS выдает заблокированные IP, а какой-то DoH незаблокированные, то это тычок пальцем в небо. Так звезды сложились в данном конкретном случае. А завтра будет другой знак зодиака, и тот же DoH выдаст заблокированные IP.

Такие вещи решаются руками. Они текучие, меняются, бегать за удобной поддержкой из коробки никто не будет.
Ставьте на саму систему DoH. В win11 и server2022 это делается в настройках. В последних win10 есть инструкции через реестр и powershell. Для остальных есть навесное 3rd party решение.

**Comment by NewUse** [comment]
now I am looking at DoT, it seems that could be very small size, I'll make some tests if it could fit to low memory routers and present results here

**Reply by bol-van** [reply]
DoT means TLS. TLS libs are huge. Http libs (libcurl) are smaller.
